pkg/receive/hashring.go

```diff
@@ -66,49 +66,106 @@ func hash(tenant string, ts *prompb.TimeSeries) uint64 {
+	// to the cache map, as this is both written to
+	// and read from.
+	mu sync.Mutex
 }
 
-// GetHost returns a hostname to handle the given tenant and time series.
-func (m matchingHashring) GetHost(tenant string, ts *prompb.TimeSeries) (string, error) {
-	if h, ok := m.cache[tenant]; ok {
-		return h.GetHost(tenant, ts)
+// Get returns a target to handle the given tenant and time series.
+func (mh *matchingHashring) Get(tenant string, ts *prompb.TimeSeries) (string, error) {
+	mh.mu.Lock()
+	defer mh.mu.Unlock()
+	if h, ok := mh.cache[tenant]; ok {
+		return h.Get(tenant, ts)
 	}
-	for name := range m.hashrings {
-		if m.matcher.Match(tenant, name) {
-			m.cache[tenant] = m.hashrings[name]
-			return m.hashrings[name].GetHost(tenant, ts)
+	var found bool
+	// If the tenant is not in the cache, then we need to check
+	// every tenant in the configuration.
+	for i, h := range mh.cfg {
+		// If the hashring has no tenants, then it is
+		// considered a default hashring and matches everything.
+		if len(h.Tenants) == 0 {
+			found = true
+		}
+		// Match against every tenant in the hashring.
+		for _, t := range h.Tenants {
+			if mh.matcher.Match(tenant, t) {
+				found = true
+				break
+			}
+		}
+		// If the hashring has no tenants, then it is
+		// considered a default hashring and matches everything.
+		if found {
+			mh.cache[tenant] = mh.hashrings[i]
+			return mh.hashrings[i].Get(tenant, ts)
 		}
 	}
-	return "", errors.New("no matching hosts to handle tenant")
+	return "", errors.New("no matching hashring to handle tenant")
 }
 
-// NewHashring creates a multi-tenant hashring for a given slice of
-// groups. Which tenant's hashring to use is determined by the Matcher.
-func NewHashring(matcher Matcher, groups []*targetgroup.Group) Hashring {
-	m := matchingHashring{
-		cache:     make(map[string]Hashring),
-		hashrings: make(map[string]Hashring),
-		matcher:   matcher,
+// newMatchingHashring creates a multi-tenant hashring for a given slice of
+// groups.
+// Which nodes are assigned to a hashring is controlled
+// by the given tenant label.
+// Which tenant's hashring to use is determined by the Matcher.
+func newMatchingHashring(matcher Matcher, cfg []HashringConfig) Hashring {
+	m := &matchingHashring{
+		cache:   make(map[string]Hashring),
+		matcher: matcher,
+		cfg:     cfg,
 	}
-	for _, g := range groups {
-		m.hashrings[g.Source] = &simpleHashring{*g}
+
+	for _, h := range cfg {
+		m.hashrings = append(m.hashrings, simpleHashring(h.Endpoints))
 	}
 	return m
 }
+
+// HashringFromConfig creates multi-tenant hashrings from a
+// hashring configuration file watcher.
+// The configuration file is watched for updates.
+// Hashrings are returned on the updates channel.
+// Which nodes are assigned to a hashring is controlled
+// by the given tenant label.
+// Which tenant's hashring to use is determined by the Matcher.
+func HashringFromConfig(ctx context.Context, updates chan<- Hashring, matcher Matcher, cw *ConfigWatcher) {
```

Comment:
The `matcher` should just be part of the configuration now, no?

Reply:
what do you mean exactly? we talked about the configuration only allowing for exact set matching, otherwise for empty set it is match all. This means we *could* remove the matcher from the func signature and always use exact match.

Reply:
Precisely, let's just always use set-matching

Reply:
Alright sounds good :) done